src/main/scala/com/fasterxml/jackson/module/scala/ser/EitherSerializer.scala

```diff
@@ -1,95 +1,174 @@
+    serialize(value, jgen, provider, None)
   }
 
-  def createContextual(prov: SerializerProvider, property: BeanProperty): JsonSerializer[_] = {
-    // Based on the version in OptionSerializer
-    def serializerFromAnnotation(property: BeanProperty, serializerProvider: SerializerProvider) = {
-      Option(property).flatMap(p => Option(p.getMember)).flatMap { m =>
-        Option(serializerProvider.getAnnotationIntrospector.findContentSerializer(m))
-          .map(serDef => serializerProvider.serializerInstance(m, serDef))
-      }
+  def serialize(value: Either[AnyRef, AnyRef], jgen: JsonGenerator, provider: SerializerProvider, vts: Option[TypeSerializer]): Unit = {
+    val (field, content, details) = value match {
+      case Left(c) => ("l", c, left)
+      case Right(c) => ("r", c, right)
     }
-    def hasContentTypeAnnotation(provider: SerializerProvider, property: BeanProperty) =
-      Option(property).exists { p =>
-        Option(provider.getAnnotationIntrospector.refineSerializationType(provider.getConfig, p.getMember, p.getType)).isDefined
-      }
-
-    def tryContentSerializer(serializerProvider: SerializerProvider, property: BeanProperty, currentSer: Option[JsonSerializer[_]]) = {
-      val ser = Option(findConvertingContentSerializer(serializerProvider, property, currentSer.orNull))
-      (ser, elementType) match {
-        case (None, Some(et)) if hasContentTypeAnnotation(serializerProvider, property) =>
-          Option(serializerProvider.findValueSerializer(et, property))
-        case (Some(sr), _) => Option(serializerProvider.handleSecondaryContextualization(sr, property))
-        case _ => ser
-      }
+    jgen.writeStartObject()
+    jgen.writeFieldName(field)
+    if (content == null) {
+      provider.defaultSerializeNull(jgen)
+      return
+    }
+    val ser = details.valueSerializer.getOrElse(findCachedSerializer(provider, content.getClass))
+    vts.orElse(details.valueTypeSerializer) match {
+      case Some(vts) => ser.serializeWithType(content, jgen, provider, vts)
+      case None => ser.serialize(content, jgen, provider)
     }
+    jgen.writeEndObject()
+  }
 
-    val typeSer = valueTypeSerializer.map(_.forProperty(property))
-    val candidateSerializer = tryContentSerializer(prov, property, serializerFromAnnotation(property, prov).orElse(elementSerializer))
+  override def serializeWithType(value: Either[AnyRef, AnyRef], jgen: JsonGenerator, provider: SerializerProvider, typeSer: TypeSerializer): Unit = {
+    if (value == null) {
+      provider.defaultSerializeNull(jgen)
+      return
+    }
 
-    if (candidateSerializer != elementSerializer || property != beanProperty.orNull || valueTypeSerializer != typeSer)
-      new EitherSerializer(elementType, typeSer, Option(property), candidateSerializer.asInstanceOf[Option[JsonSerializer[AnyRef]]])
-    else this
+    // Otherwise apply type-prefix/suffix, then std serialize:
```

Comment:
Nice, I like simplicity here, esp. compared to `Optional` which has to work around the fact that only one of `TypeSerializer`s is needed (since it does not add JSON Array or Object around value).